### PR TITLE
[WIP] Retrieve() modes.

### LINKED ISSUE #1195: Provide path for Bob to create new WorkOrder for same Ursula for same Capsule
#874 empowered Bob to retrieve the same secret on multiple occasions by optionally caching `CFrags` (in fact, an entire activated `Capsule`).

### PR DESCRIPTION
**Pull request by jMyles**
* Fixes #1195

### CODE REVIEW

**Comment by cygnusv** [comment]
What about also addressing these issues on this PR?

https://github.com/nucypher/nucypher/issues/1114
https://github.com/nucypher/nucypher/issues/1043

**Comment by jMyles** [comment]
@cygnusv: Both had occurred to me.  I'm actually working on the latter as we speak (from the corner of a cozy little bar just outside Amsterdam Centrum).

However, if I'm really telling me truth: I'd like to either get it merged while the tests are passing (and open subsequent PRs for these matters) or turn this into an epic.

Are you interested in working on either of these?

File: nucypher/characters/lawful.py [hunk 1]
```diff
@@ -511,7 +509,7 @@ def _pick_treasure_map(self, treasure_map=None, map_id=None):
             else:
                 raise ValueError("You need to pass either treasure_map or map_id.")
         elif map_id:
-                raise ValueError("Don't pass both treasure_map and map_id - pick one or the other.")
+            raise ValueError("Don't pass both treasure_map and map_id - pick one or the other.")
```
Comment (KPrasch) [inline nucypher/characters/lawful.py]: We need to form a convention for this language - Perhaps even an exception class.

Comment (cygnusv) [reply nucypher/characters/lawful.py]: You mean for this XOR-like argument passing?

File: nucypher/characters/lawful.py [hunk 2]
```diff
@@ -640,7 +638,13 @@ def get_treasure_map_from_known_ursulas(self, network_middleware, map_id):
 
         return treasure_map
 
-    def generate_work_orders(self, map_id, *capsules, num_ursulas=None, cache=False):
+    def work_orders_for_capsules(self,
+                                 *capsules,
```
Comment (KPrasch) [inline nucypher/characters/lawful.py]: Let's use "forced kwargs" here `*`.

File: tests/characters/test_bob_handles_frags.py [hunk 3]
```diff
@@ -311,24 +346,296 @@ def test_bob_gathers_and_combines(enacted_federated_policy, federated_bob, feder
...
+
+    delivered_cleartexts = federated_bob.retrieve(the_message_kit,
+                                                  enrico=capsule_side_channel.enrico,
+                                                  alice_verifying_key=alices_verifying_key,
+                                                  label=enacted_federated_policy.label,
+                                                  retain_cfrags=True)
+    assert b"Welcome to flippering number 1." == delivered_cleartexts[0]
+
+    # Can't retrieve this message again.
+    # Bob needs to either instantiate the message_kit again or use use_attached_cfrags=True.
```
Comment (michwill) [inline tests/characters/test_bob_handles_frags.py]: So, we're kinda encouraging to not store sensitive materials this way?

Comment (jMyles) [reply tests/characters/test_bob_handles_frags.py]: Right - not without expressly declaring that intent.  

It's not per se a worse way to store sensitive materials than any other way, it just has a different balance of concerns, which we need to clarify in the docs.

File: nucypher/policy/models.py [hunk 4]
```diff
@@ -632,15 +632,12 @@ def construct_by_bob(cls, arrangement_id, capsules, ursula, bob):
...
             ursula_identity_evidence = ursula.decentralized_identity_evidence
 
-        tasks, tasks_bytes = [], []
+        tasks, tasks_bytes = {}, []
         for capsule in capsules:
-            if alice_verifying_key != capsule.get_correctness_keys()["verifying"]:
-                raise ValueError("Capsules in this work order are inconsistent.")
-
-            task = cls.Task(capsule, signature=None)
+            task = cls.PRETask(capsule, signature=None)
```
[the remark this answers is not in the document]
Comment (cygnusv) [reply nucypher/policy/models.py]: That's a poor design and it's my fault. `Task` (now `PRETask`) is just intended to be little more than a data structure to store the capsule and some additional stuff. Initially, one just have the capsule and later on the signature can be added. Perhaps the best way forward is to make the signature optional.

File: tests/cli/test_bob.py [hunk 5]
```diff
@@ -8,3 +18,77 @@ def test_bob_cannot_init_with_dev_flag(click_runner):
...
+                     '--config-file', bob_configuration_file_location,
+                     '--message-kit', message_kit_b64_bytes,
+                     '--label', label,
+                     '--policy-encrypting-key', federated_alice.get_policy_encrypting_key_from_label(label).hex(),
+                     '--alice-verifying-key', federated_alice.public_keys(SigningPower).hex()
+                     )
+
+    from nucypher.cli import actions
+
+    def substitue_bob(*args, **kwargs):
```
Comment (mswilkison) [inline tests/cli/test_bob.py]:
```suggestion
    def substitute_bob(*args, **kwargs):
```

File: tests/cli/test_bob.py [hunk 6]
```diff
@@ -8,3 +18,77 @@ def test_bob_cannot_init_with_dev_flag(click_runner):
...
+        this_fuckin_guy = enacted_federated_policy.bob
+        somebody_else = Ursula.from_teacher_uri(teacher_uri=kwargs['teacher_uri'],
+                                               min_stake=0,
+                                               federated_only=True,
+                                               network_middleware=this_fuckin_guy.network_middleware)
+        this_fuckin_guy.remember_node(somebody_else)
+        this_fuckin_guy.controller.emitter = JSONRPCStdoutEmitter()
+        return this_fuckin_guy
+
+    actions.make_cli_character = substitue_bob
```
Comment (mswilkison) [inline tests/cli/test_bob.py]:
```suggestion
    actions.make_cli_character = substitute_bob
```

**Pull request merged**